src/gurt/telemetry.c

```diff
@@ -1824,6 +1853,269 @@ d_tm_add_metric(struct d_tm_node_t **node, char *metric, int metric_type,
+
+	dth_buckets = metric->dtm_histogram->dth_buckets;
+
+	min = 0;
+	max = initial_width - 1;
+	for (i = 0; i < num_buckets; i++) {
+		D_ASPRINTF(meta_data, "histogram bucket %d [%lu .. %lu]",
+			   i, min, max);
+		if (meta_data == NULL) {
+			rc = -DER_NOMEM;
+			goto failure;
+		}
+
+		D_ASPRINTF(fullpath, "%s/bucket %d", path, i);
+		if (fullpath == NULL) {
+			rc = -DER_NOMEM;
+			goto failure;
+		}
+
+		dth_buckets[i].dtb_min = min;
+		dth_buckets[i].dtb_max = max;
+
+		rc = d_tm_add_metric(&dth_buckets[i].dtb_bucket, fullpath,
+				     D_TM_COUNTER, meta_data, "");
+		D_FREE(fullpath);
+		D_FREE(meta_data);
+		if (rc)
+			goto failure;
+
+		min = max + 1;
+
+		if (i == (num_buckets - 2)) {
+			max = UINT64_MAX;
+		} else if (multiplier == 1) {
+			max += initial_width;
+		} else {
+			max = (min * multiplier) - 1;
+		}
+	}
+
+	D_DEBUG(DB_TRACE, "Successfully added histogram for: [%s]\n", path);
+	return D_TM_SUCCESS;
+
+failure:
+
+	D_ERROR("Failed to histogram for [%s]: " DF_RC "\n", path, DP_RC(rc));
+	return rc;
+}
+
+/**
+ * Retrieves the histogram creation data for the given node, which includes
+ * the number of buckets, initial width and multiplier used to create the
+ * given histogram.
+ *
+ * \param[in,out]	histogram	Pointer to a d_tm_histogram_t used to
+ *					store the results.
+ * \param[in]		shmem_root	Pointer to the shared memory segment.
+ * \param[in]		node		Pointer to the metric node with a
+ *					histogram.
+ *
+ * \return			D_TM_SUCCESS		Success
+ *				-DER_INVAL		node or histogram is
+ *							invalid.
+ *				-DER_METRIC_NOT_FOUND	The metric node, the
+ *							metric data or histogram
+ *							was not found.
+ *				-DER_OP_NOT_PERMITTED	Node was not a gauge
+ *							or duration with
+ *							an associated histogram.
+ */
+int
+d_tm_get_num_buckets(struct d_tm_histogram_t *histogram,
+		     uint64_t *shmem_root, struct d_tm_node_t *node)
+{
+	struct d_tm_histogram_t	*dtm_histogram = NULL;
+	struct d_tm_metric_t	*metric_data = NULL;
+
+	if (node == NULL)
+		return -DER_INVAL;
+
+	if (histogram == NULL)
+		return -DER_INVAL;
+
+	if (!d_tm_validate_shmem_ptr(shmem_root, (void *)node))
+		return -DER_METRIC_NOT_FOUND;
+
+	if (!((node->dtn_type == D_TM_GAUGE) ||
+	      (node->dtn_type & D_TM_DURATION)))
+		return -DER_OP_NOT_PERMITTED;
+
+	metric_data = d_tm_conv_ptr(shmem_root, node->dtn_metric);
+	if (metric_data == NULL)
+		return -DER_METRIC_NOT_FOUND;
+
+	dtm_histogram = d_tm_conv_ptr(shmem_root, metric_data->dtm_histogram);
+	if (dtm_histogram == NULL)
+		return -DER_METRIC_NOT_FOUND;
+
+	histogram->dth_num_buckets = dtm_histogram->dth_num_buckets;
+	histogram->dth_initial_width = dtm_histogram->dth_initial_width;
+	histogram->dth_value_multiplier = dtm_histogram->dth_value_multiplier;
+
+	return D_TM_SUCCESS;
+}
+
+/**
+ * Retrieves the range of the given bucket for the node with a histogram.
+ *
+ * \param[in,out]	histogram	Pointer to a d_tm_histogram_t used to
+ *					store the results.
```

Comment:
Not accurate here? Looks like it's a bucket instead, in the function param list

Reply:
Ooops.  Good catch.